### PR TITLE
fix(tools): align serper, firecrawl, langsmith and qdrant declared shapes with the APIs

### PR DESCRIPTION
**Pull request by waleedlatif1**
Declared tool outputs that do not match what the provider actually returns. Every shape change below is backed by a primary doc, cited inline. Each fix has a test that was verified to fail against the old code first.

## 1. Serper `places` mapped fields `/places` does not return

`places.toResult` emitted `link`, `snippet` and `reviews`. A `/places` item has none of those keys, so all three resolved to `undefined` on every result, while the fields Serper does return (`ratingCount`, `category`, `phoneNumber`, `website`, `latitude`, `longitude`, `cid`) were dropped. `ratingCount` was a **declared** output that no code path could ever populate — the review count arrives under that name, not as `reviews`.

Documented `/places` item (serper.dev, via the published response examples):

```json
{ "position": 1, "title": "Whole Foods Market", "address": "2001 Market St, San Francisco, CA 94114",
  "latitude": 37.7687616, "longitude": -122.42701559999999, "rating": 4.2, "ratingCount": 1500,
  "category": "Grocery store", "phoneNumber": "(415) 626-1430", "website": "https://…", "cid": "6353588238324409422" }
```

`link` and `snippet` stay **declared but are now `optional`** — they are real on every other vertical, so they are documented as absent on places rather than removed. `reviews` was never a declared output (only a TS field commented "not an advertised output"), so dropping it removes no advertised surface.

Also: `news` items always carry `source` (declared on the unified result as "Source name (news/videos/shopping)") and the mapper dropped it. Now mapped.

Docs: <https://serper.dev> (places + news response examples). Cross-checked against the `News` type in the community client at <https://github.com/tkdkid1000/serper/blob/main/docs/types/News.html>.

## 2. Serper `/search` panels were undeclared

`peopleAlsoAsk` and `relatedSearches` are top-level keys on the `/search` payload and were never surfaced. Both are now declared and populated, and stay absent (not empty arrays) on verticals that do not return them.

Docs: <https://serper.dev> (search response example showing both blocks verbatim).

`knowledgeGraph` is **not** included — see "Unverifiable" below.

## 3. Firecrawl search declared the wrong container

`SearchResponse.output.data` was `Array<…>`. `POST /v2/search` answers with a **source-keyed envelope**: `data.web`, `data.news`, `data.images`, present only for the requested sources. News items carry `snippet` (not `description`); image items put the containing page in `url` and the image in `imageUrl`. Declared outputs now model all three item shapes, and an absent body resolves to `{}` rather than `undefined`.

Docs: <https://docs.firecrawl.dev/api-reference/v2-openapi.json> (`paths./search.post.responses.200`). Cross-checked against `SearchData` / `SearchResultWeb` / `SearchResultNews` / `SearchResultImages` in the official JS SDK: <https://github.com/firecrawl/firecrawl/blob/main/apps/js-sdk/firecrawl/src/v2/types.ts>.

Note the drift's origin: the repo's `apps/api/openapi.json` and `v1-openapi.json` both still describe **v1**, where `data` genuinely was an array. The tool calls `/v2`.

## 4. Firecrawl `map` — `timeout` param renamed, and `links` are objects

`request-transport.ts:191` reads `params.timeout` as the outbound fetch deadline for **every** tool. `firecrawl_map` declared a param of that name intended for Firecrawl's own map deadline, so setting it also armed the local abort at the same instant — the client abort wins the race and the caller gets an opaque `AbortError` instead of Firecrawl's response. Renamed to `mapTimeout` and mapped to the body's `timeout`.

**No subBlock id changed.** The block's `timeout` subBlock is conditioned on `['scrape', 'search', 'parse']` and the `map` branch of `tools.config.params` never forwarded it, so no saved workflow state referenced this param — the rename is confined to the tool's direct-call surface. Nothing needed adding to `tools.config.params`.

`firecrawl/parse.ts`'s `timeout` is left alone as instructed, and is unchanged in this PR.

Separately, `/v2/map` returns `links` as `SearchResultWeb[]` — objects with a required `url` plus optional `title`/`description` — while the tool declared `items: { type: 'string' }`. Corrected.

Docs: <https://docs.firecrawl.dev/api-reference/v2-openapi.json> (`components.schemas.MapResponse`, and `paths./map.post.requestBody` for `timeout`: "Timeout in milliseconds. There is no timeout by default."). Cross-checked against `MapData.links: SearchResultWeb[]` in the official JS SDK.

## 5. LangSmith feedback score coercion only ran in the block

`blocks/blocks/langsmith.ts` did the `Number(value)` + NaN check inside `tools.config.params`, so only the block path got it — LLM tool calls and direct tool calls posted the raw value. Moved into `create_feedback.ts`'s `body`, so every caller is covered; the block now passes the value straight through. Behavior is preserved exactly, including the `Invalid score: "…" is not a number` message and treating `''`/`null`/`undefined` as omitted.

## 6. Qdrant block declared outputs no tool emits

`QdrantBlock.outputs` declared `matches` and `upsertedCount`. All three Qdrant tools return exactly `{ status, data }` — no code path has ever produced either key, so every reference resolved to `undefined`. Both removed; `data`'s description now says what it holds per operation.

This one needed no API doc — it is provable from the repo. For completeness the point shapes were checked against Qdrant's OpenAPI (`Record`, `ScoredPoint`, `UpdateResult` at <https://github.com/qdrant/qdrant/blob/master/docs/redoc/master/openapi.json>) and every declared per-point field (`version`, `score`, `shard_key`, `order_value`, `operation_id`) is real — **no phantom fields found in the qdrant tool layer**.

## Fields removed, and the evidence

| Field | Where | Evidence it is never returned |
|---|---|---|
| `reviews` | serper places mapper + `SearchResult` | `/places` items report the review count as `ratingCount`. `reviews` appears in no serper.dev example. Was not a declared output. |
| `matches` | `QdrantBlock.outputs` | No Qdrant `transformResponse` writes it; all three return `{ status, data }`. |
| `upsertedCount` | `QdrantBlock.outputs` | Same. Qdrant's `UpdateResult` has only `operation_id` and `status`. |

`link` and `snippet` on places results are **narrowed to optional, not removed** — they remain populated on every other vertical.

## Unverifiable

- **Serper `knowledgeGraph`.** `SearchResponse` has carried a dead `knowledgeGraph?` TS field and a `KNOWLEDGE_GRAPH_OUTPUT_PROPERTIES` constant, neither declared as an output nor populated. I could not retrieve a primary serper.dev example of the object, so I did not add it. To finish it I need a captured `/search` payload for an entity query (e.g. `q: "OpenAI"`) or a live Serper key. Left exactly as-is.
- Same for `answerBox` and `topStories`: dead TS-only fields, no advertised output, untouched.

## Tests

Every assertion was run against the unmodified source first and confirmed red (13 failures), then green after the fix, then the source changes were stashed and the same 13 confirmed red again.

- `tools/serper/search.test.ts` — places payload copied from serper.dev; asserts no `link`/`snippet`/`reviews`; news `source`; the two `/search` panels present on `search` and absent on `news`.
- `tools/firecrawl/search.test.ts` (new) — source-keyed envelope, news `snippet` not `description`, image `url` vs `imageUrl`, empty-envelope fallback, declared-output shape, `mapTimeout` rename and body mapping, map link objects.
- `tools/langsmith/langsmith.test.ts` — string coercion, `0` preserved, blank omitted, non-numeric rejected, all at the tool layer.
- `tools/qdrant/search_vector.test.ts` — block declares only `data` and `status`; no tool emits `matches`/`upsertedCount`.

## Gates

`tool-metadata:generate`, `generate-docs.ts` (generated diff committed), `lint`, `check:audits` (39/39, includes `docs:check`), `check-block-registry.ts` all pass. Type-check is clean for these files; the two `@aws-sdk/client-lambda` errors are pre-existing and unrelated.

### CODE REVIEW

File: apps/sim/tools/langsmith/create_feedback.ts [hunk 1]
```diff
@@ -7,6 +7,20 @@ import type {
...
 
+/**
+ * LangSmith stores `score` as a number. The value reaches this tool as a string from the block's
+ * short-input, as a number from a direct tool call, and as either from an LLM tool call, so the
+ * coercion lives here rather than in any one caller.
+ */
+function coerceFeedbackScore(value: unknown): number | undefined {
+  if (value === undefined || value === null || value === '') return undefined
+  const parsed = Number(value)
+  if (Number.isNaN(parsed)) {
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/langsmith/create_feedback.ts]: Valid — fixed in 5a7d035b. `Number("Infinity")` is `Infinity`, which passes `Number.isNaN`, and `JSON.stringify(Infinity)` is `null`, so LangSmith would have stored a blank score. Now guarded with `Number.isFinite`, which covers NaN and both infinities in one check.

Worth recording where this came from: the whole `coerceFeedbackScore` helper is a **verbatim move** of the `parseScore` closure that lived at `blocks/blocks/langsmith.ts:392` before this PR — same `Number(value)`, same `Number.isNaN` guard. So this is a pre-existing bug I inherited by relocating the code, not one this PR introduced. Moving it to the tool layer is what made it reachable from the LLM and direct-tool paths too, which is a good reason to fix it here rather than defer.

Covered by `langsmith create feedback score edge cases > rejects the non-finite score ...` in `tools/langsmith/langsmith.test.ts`, parameterised over `"Infinity"`, `"-Infinity"`, and the numeric forms. Confirmed red before the fix.

File: apps/sim/tools/langsmith/create_feedback.ts [hunk 2]
```diff
@@ -7,6 +7,20 @@ import type {
 import { LANGSMITH_API_BASE, truncateLangsmithErrorText } from '@/tools/langsmith/utils'
 import type { ToolConfig } from '@/tools/types'
 
+/**
+ * LangSmith stores `score` as a number. The value reaches this tool as a string from the block's
+ * short-input, as a number from a direct tool call, and as either from an LLM tool call, so the
+ * coercion lives here rather than in any one caller.
+ */
+function coerceFeedbackScore(value: unknown): number | undefined {
+  if (value === undefined || value === null || value === '') return undefined
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/langsmith/create_feedback.ts]: Valid — fixed in 5a7d035b. `Number("   ")` is `0`, so a short-input containing only whitespace recorded a real score of zero instead of omitting a blank field. Now `typeof value === "string" && value.trim() === ""` returns `undefined` before coercion.

Same provenance note as the sibling thread: this is the `parseScore` closure moved verbatim from `blocks/blocks/langsmith.ts:392`, which had the identical `value === ""` check that a whitespace string slips past. Inherited, not introduced — but fixed here since the move is what put it on every caller path.

I deliberately kept `0` itself meaningful: `score: 0` is a legitimate LangSmith score and still posts. Test `omits a whitespace-only score instead of recording it as 0` covers the blank case, and the pre-existing `keeps a number score untouched, zero included` pins the other side.

File: apps/sim/tools/firecrawl/types.ts [hunk 3]
```diff
@@ -346,25 +445,63 @@ export interface ScrapeResponse extends ToolResponse {
...
+  title?: string
+  description?: string
+  sourceURL: string
+  url?: string
+  statusCode?: number
+  error?: string | null
+}
+
+interface ScrapedSearchContent {
+  markdown?: string | null
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/firecrawl/types.ts:534]: Valid — fixed in 5a7d035b, and I re-checked the two sources this PR already cites rather than taking the finding on faith.

The v2 schema marks these explicitly nullable on **both** `data.web[]` and `data.news[]` items:

```
web nullable fields:  markdown, html, rawHtml, screenshot, audio, video
news nullable fields: markdown, html, rawHtml, screenshot, audio, video
metadata.error nullable: true
```

(read out of `paths./search.post.responses.200` in <https://docs.firecrawl.dev/api-reference/v2-openapi.json>)

`OutputProperty` already has a `nullable?: boolean` slot, so the advertised schema can say this precisely instead of me normalising nulls away — normalising would discard the distinction between "not requested" and "requested but empty", which is exactly the kind of lossy declaration this PR is trying to remove. `markdown`, `html`, `rawHtml` and `screenshot` are now `optional: true, nullable: true` on both item sets, and `metadata.error` likewise.

I did not touch `audio`/`video`: they are real in the schema but this tool never requests those formats, and adding outputs the tool cannot produce would reintroduce the phantom-output defect the PR exists to fix.

Test: `firecrawl_search declared nullability > marks scraped content nullable and metadata optional on web and news items`. Confirmed red first.

File: apps/sim/tools/firecrawl/types.ts [hunk 4]
```diff
@@ -168,13 +170,105 @@ export const SEARCH_RESULT_OUTPUT_PROPERTIES = {
...
+    description: 'Links found on the article page; returned only when "links" was requested',
+    optional: true,
+    items: { type: 'string', description: 'URL found on the page' },
+  },
+  screenshot: {
+    type: 'string',
+    description: 'Screenshot URL (expires after 24 hours); returned only when requested',
+    optional: true,
+  },
+  metadata: SEARCH_METADATA_OUTPUT,
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/firecrawl/types.ts:286]: Valid — fixed in 5a7d035b. Two independent confirmations, both from sources already cited in the PR body:

1. The v2 schema declares **no `required` list at all** on either the `data.web[]` or `data.news[]` item objects, so nothing on them is guaranteed — `metadata` included.
2. The official SDK is more explicit: `SearchData.web?: Array<SearchResultWeb | Document>`, and `SearchResultWeb` has **no `metadata` field whatsoever**. Metadata only appears on the `Document` variant, i.e. the scraped one. So an unscraped hit genuinely arrives without it. (<https://github.com/firecrawl/firecrawl/blob/main/apps/js-sdk/firecrawl/src/v2/types.ts>)

`SEARCH_METADATA_OUTPUT` is now `optional: true` with a description saying it is present only when the result was scraped, which is shared by both item sets.

The same SDK reading made me widen slightly further than you asked, and I want to flag it rather than slip it in: `SearchResultWeb` types `url` as required but `title`/`description` as optional, and `SearchResultNews` types *every* field optional. I have matched that on both the TS interfaces and the declared outputs. Declaring `title` required when the SDK says otherwise is the same class of defect as the original `data: Array<...>` — so I would rather be accurate than minimal here. Happy to narrow it back if you read the SDK differently.

File: apps/sim/tools/firecrawl/map.ts [hunk 5]
```diff
@@ -88,7 +90,7 @@ export const mapTool: ToolConfig<MapParams, MapResponse> = {
       if (typeof params.ignoreQueryParameters === 'boolean')
         body.ignoreQueryParameters = params.ignoreQueryParameters
       if (params.limit) body.limit = Number(params.limit)
-      if (params.timeout) body.timeout = Number(params.timeout)
+      if (params.mapTimeout) body.timeout = Number(params.mapTimeout)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/firecrawl/map.ts]: Valid — fixed in 5a7d035b, and checking it turned up a second bug in the same file that I have also fixed.

On the zero itself: I could not find server-side semantics for `timeout: 0` documented anywhere — the v2 schema says only `{"type":"integer","description":"Timeout in milliseconds. There is no timeout by default."}`, with no `minimum` and no note on zero. So I am not claiming `0` is meaningful to Firecrawl. What settled it is that the **official SDK forwards it**: `prepareMapPayload` guards with `if (options.timeout != null) payload.timeout = options.timeout`, and unlike `prepareSearchPayload` (which throws `"timeout must be positive"`) the map path has no positivity check at all. Silently substituting the provider default for a value the user explicitly set is the tool lying about what it sent, so I now transmit it faithfully and let Firecrawl decide.

Implemented as a small `resolveMapTimeout` helper rather than the suggested `|| === 0`, because a blank short-input arrives as `""` and `"" != null` is true — the `||`-form would have started sending `timeout: 0` for every user who left the field empty, which is a worse bug than the one being fixed. The helper drops `undefined`/`null`/whitespace-only and keeps any finite number including `0`.

**Second finding from the same file:** `map.ts` in the SDK normalises `links?: Array<string | SearchResultWeb>` — `if (typeof item === "string") links.push({ url: item })`. So `/v2/map` can still answer with bare URL strings, which would have violated the object shape I newly declared. `transformResponse` now widens strings to `{ url }` the same way. Thanks — the finding is what sent me back to that file.

Tests: `sends an explicit mapTimeout of 0 ...`, `omits the timeout for a blank mapTimeout`, `widens a bare URL string link into the declared object shape`. All confirmed red first.

File: apps/sim/blocks/blocks/qdrant.ts [hunk 6]
```diff
@@ -257,9 +257,11 @@ Return ONLY the JSON object.`,
   },
 
   outputs: {
-    matches: { type: 'json', description: 'Search matches' },
-    upsertedCount: { type: 'number', description: 'Upserted count' },
-    data: { type: 'json', description: 'Response data' },
+    data: {
+      type: 'json',
+      description:
+        'Result payload: matched points for search, fetched points for fetch, or the update result for upsert',
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/blocks/blocks/qdrant.ts:263]: Valid, and a fair catch on my own change — fixed in 5a7d035b.

To be precise about the history: the skill was already wrong before this PR. Qdrant's `UpdateResult` has exactly two properties, `operation_id` and `status` (<https://github.com/qdrant/qdrant/blob/master/docs/redoc/master/openapi.json>) — there has never been an upserted count on the wire, and the `upsertedCount` block output I removed was itself a phantom that no `transformResponse` ever wrote. So the skill has always been instructing agents to report a number that did not exist. Removing the output is what made the inconsistency visible, which is the right outcome.

The skill now reads:

> 3. Confirm the operation succeeded from `status` and read `data.operation_id` and `data.status` in the response.
>
> **Output** — Report the operation id and its status (`acknowledged` or `completed`) for the collection written to, and surface any payload validation issues. Qdrant does not return a count of upserted points, so report the number of points that were sent instead of inventing one.

I added that last clause deliberately: without it an agent asked "how many points landed?" will reach for the nearest plausible number. Telling it to report the input count, explicitly labelled as such, is the honest answer available.

File: apps/docs/content/docs/integrations/firecrawl.mdx [hunk 7]
```diff
@@ -163,21 +163,48 @@ Search for information on the web using Firecrawl
...
-|   ↳ `rawHtml` | string | Unprocessed raw HTML; returned only when "rawHtml" is among the scrape formats requested via the hidden scrapeOptions input |
-|   ↳ `links` | array | Links found on the page; returned only when "links" is among the scrape formats requested via the hidden scrapeOptions input |
-|   ↳ `screenshot` | string | Screenshot URL \(expires after 24 hours\); returned only when "screenshot" is among the scrape formats requested via the hidden scrapeOptions input |
-|   ↳ `metadata` | object | Metadata about the search result page |
-|     ↳ `title` | string | Page title |
-|     ↳ `description` | string | Page meta description |
-|     ↳ `sourceURL` | string | Original source URL |
-|     ↳ `statusCode` | number | HTTP status code |
-|     ↳ `error` | string | Error message if scrape failed |
+| `data` | object | Search results keyed by source \(web, news, images\), each with optional scraped content and metadata |
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/docs/content/docs/integrations/firecrawl.mdx:168]: Valid, and I went with the first option — exposed `sources` rather than scoping the docs back. Fixed in 5a7d035b.

You are right that this was unreachable: `firecrawl_search.params` declared only `query`, `scrapeOptions` and `apiKey`, so no user, block field, or LLM tool call could ask for `news` or `images`, even though the request body builder already had `if (params.sources) body.sources = params.sources` sitting there unused. Documenting result arrays nobody can request is precisely the defect class this PR set out to fix, so deferring it would have been incoherent.

What changed:

- `firecrawl_search` now declares `sources` (`array`, `user-or-llm`) and `limit` (`number`, `user-or-llm`). `limit` was in the same state — supported by the body builder, undeclared, unreachable.
- The Firecrawl block gets a multi-select `sources` dropdown (Web / News / Images) conditioned on `operation === "search"`, forwarded through `tools.config.params` via a `parseSearchSources` helper that tolerates both the array form and the JSON-string form an older saved revision could hold, and drops unknown entries rather than forwarding a source Firecrawl would reject.
- The regenerated `firecrawl.mdx` now lists `sources` as an input, so the documented `data.news` / `data.images` arrays are backed by a real control.

The bare-string form (`["web","news"]`) is what I send, matching `SearchRequest.sources?: Array<"web" | "news" | "images" | { type: ... }>` in the official SDK — the schema also accepts the object form, but the string form is simpler and equally valid.

`check-block-registry.ts` passes including the subblock ID stability check — `sources` is a new id, purely additive, and no existing id moved. Tests: `firecrawl_search sources > exposes sources ...`, `forwards the requested sources to Firecrawl`, `omits sources entirely when none were chosen, letting Firecrawl default to web`. Confirmed red first.

**Comment by waleedlatif1** [comment]
@cubic-dev-ai review

File: apps/sim/tools/firecrawl/types.ts [hunk 8]
```diff
@@ -163,18 +186,119 @@ export const SEARCH_RESULT_OUTPUT_PROPERTIES = {
...
+    optional: true,
+    items: { type: 'string', description: 'URL found on the page' },
+  },
+  screenshot: {
+    type: 'string',
+    description: 'Screenshot URL (expires after 24 hours); returned only when requested',
+    optional: true,
+    nullable: true,
+  },
+  metadata: SEARCH_METADATA_OUTPUT,
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/firecrawl/types.ts:286]: You are right and I was wrong — fixed in c0f786e3.

I explicitly declined this one round ago, on the thread about nullability, with the reasoning "this tool never requests those formats, and adding outputs the tool cannot produce would reintroduce the phantom-output defect the PR exists to fix." That reasoning does not survive checking. `scrapeOptions` is a `hidden` param described as "Advanced scrape options supplied by existing configurations", its `formats` slot is typed `FirecrawlFormat[]` — i.e. free-form `string | { type: string, ... }` — and `search.ts` forwards it verbatim with `body.scrapeOptions = params.scrapeOptions`. So a saved configuration absolutely can request `audio` or `video`, and Firecrawl will answer with those signed URLs. They are reachable, which makes them the opposite of phantom.

`audio` and `video` are now declared `optional: true, nullable: true` on both `data.web[]` and `data.news[]` (the v2 schema marks both nullable), and added to the shared `ScrapedSearchContent` interface. Descriptions note the 1-hour signed-URL expiry.

Test: `declares nullable audio and video on web/news items`. Confirmed red first.

File: apps/sim/tools/firecrawl/types.ts [hunk 9]
```diff
@@ -346,25 +475,63 @@ export interface ScrapeResponse extends ToolResponse {
   }
 }
 
+interface SearchResultMetadata {
+  title?: string
+  description?: string
+  sourceURL: string
+  url?: string
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/firecrawl/types.ts:528]: Valid, and it is an inconsistency I introduced — fixed in c0f786e3.

I added `url?: string` to the `SearchResultMetadata` TypeScript interface last round but never mirrored it into `SEARCH_METADATA_OUTPUT_PROPERTIES`, so the declared output disagreed with the type in the same file. Exactly the drift this PR is about, committed by me. The v2 schema documents it as "The final URL of the page after all redirects have been followed", distinct from `sourceURL` ("The original URL that was requested").

Both are now declared, and I reworded `sourceURL` to say it differs from `url` when a redirect occurred, so consumers picking between them have the distinction in front of them rather than having to infer it.

Test: `declares metadata.url, the final URL after redirects`, which also pins `sourceURL` as still required.

File: apps/sim/tools/firecrawl/types.ts [hunk 10]
```diff
@@ -88,15 +88,27 @@ export const SEARCH_METADATA_OUTPUT_PROPERTIES = {
   description: { type: 'string', description: 'Page meta description', optional: true },
   sourceURL: { type: 'string', description: 'Original source URL' },
   statusCode: { type: 'number', description: 'HTTP status code', optional: true },
-  error: { type: 'string', description: 'Error message if scrape failed', optional: true },
+  error: {
+    type: 'string',
+    description: 'Error message if scrape failed',
+    optional: true,
+    nullable: true,
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/firecrawl/types.ts:104]: Valid — and this one was an accident on my part, now reverted in c0f786e3.

The nullability pass in the previous commit was applied with a regex over the field names `markdown`/`html`/`rawHtml`/`screenshot`, and it matched more than the two search item sets I had actually verified: it also landed `nullable: true` on `CRAWLED_PAGE_OUTPUT_PROPERTIES.screenshot`. That is a declaration about the **crawl** response, which this PR has not checked against the v2 schema, and it contradicted `CrawledPage.screenshot?: string` in the same file — so I shipped exactly the tool/type disagreement the PR is meant to remove.

Rather than resolve it the way you suggested (widening `CrawledPage.screenshot` to `string | null`), I have reverted the declaration back to `optional: true` with no `nullable`. Reason: I have no verified evidence about crawl-response nullability. The v2 crawl schema may well mark it nullable, and if so that is a real follow-up — but making the *type* match an assertion I introduced by accident would be cementing an unverified claim rather than correcting one. Crawl outputs are now byte-identical to `origin/staging`, which is where an unreviewed area belongs in this PR.

Added a regression test asserting `CRAWLED_PAGE_OUTPUT_PROPERTIES.screenshot` carries no `nullable`, so a future broad-brush edit into crawl territory fails loudly. Confirmed red against the previous commit.

File: apps/docs/content/docs/integrations/firecrawl.mdx [hunk 11]
```diff
@@ -282,15 +311,18 @@ Get a complete list of URLs from any website quickly and reliably. Useful for di
 | `includeSubdomains` | boolean | No | Whether to include URLs from subdomains \(default: true\) |
 | `ignoreQueryParameters` | boolean | No | Exclude URLs containing query strings \(default: true\) |
 | `limit` | number | No | Maximum number of links to return \(e.g., 100, 1000, 5000\). Max: 100,000, default: 5,000 |
-| `timeout` | number | No | Request timeout in milliseconds |
+| `mapTimeout` | number | No | Firecrawl's own map deadline in milliseconds, sent as the request body's `timeout`. Named `mapTimeout` because a tool param called `timeout` is consumed as the outbound fetch deadline instead of reaching Firecrawl. |
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/docs/content/docs/integrations/firecrawl.mdx:322]: Valid, and consistent with how I just handled `sources` — exposed rather than de-documented. Fixed in c0f786e3.

You are right that the gap predates the rename in a sense: the block's shared `timeout` subblock is conditioned on `["scrape", "search", "parse"]` and the `map` branch of `tools.config.params` never forwarded it, so a map-side deadline was unreachable from the block before this PR too. What changed is that the tool param is now called `mapTimeout` and appears under that name in the generated `firecrawl.mdx`, so the unreachability is newly visible in the docs. I called this out in the PR body but then left it — that was the wrong call, and it is the same "documented capability with no input path" defect I opened the PR to fix.

- New `mapTimeout` subblock (`short-input`, "Map Timeout (ms)"), conditioned on `operation === "map"` only.
- Forwarded from the `map` branch. Note it is passed **raw**, not through `Number.parseInt`: the tool's own `resolveMapTimeout` already keeps an explicit `0` and drops a blank field, whereas `Number.parseInt("")` yields `NaN` and would have re-broken the zero-preservation fix from the sibling thread.
- Declared as a block input so the generated docs describe something reachable.

I deliberately did **not** widen the existing shared `timeout` subblock to cover `map`. That field feeds `params.timeout`, which `request-transport.ts` consumes as the outbound fetch deadline — routing map through it would restore precisely the hijack this PR removes. A separate id is the point, not an inconvenience.

`check-block-registry.ts` passes including subblock ID stability: `mapTimeout` is a new id, purely additive, no existing id moved or renamed, so no saved workflow state is orphaned.

**Comment by waleedlatif1** [comment]
@cubic-dev-ai review

File: apps/sim/tools/langsmith/create_feedback.ts [hunk 12]
```diff
@@ -7,6 +7,25 @@ import type {
...
 
+/**
+ * LangSmith stores `score` as a number. The value reaches this tool as a string from the block's
+ * short-input, as a number from a direct tool call, and as either from an LLM tool call, so the
+ * coercion lives here rather than in any one caller.
+ */
+function coerceFeedbackScore(value: unknown): number | undefined {
+  if (value === undefined || value === null) return undefined
+  if (typeof value === 'string' && value.trim() === '') return undefined
+  const parsed = Number(value)
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/langsmith/create_feedback.ts:26]: Valid — fixed in 163f2f4e. `Number(false)` is `0` and `Number([])` is `0`, both of which sail through `Number.isFinite`, so a malformed value from an LLM or a direct tool call was being written to LangSmith as a real evaluation score of zero. `Number(true)` is `1`, which is worse — a plausible-looking score with no basis at all.

Coercion is now restricted to `number | string` up front; anything else throws the same `Invalid score: "..." is not a number`. `String(value)` is used to build the message so an object or array cannot blow up the interpolation.

Ordering matters here and I put the type guard **before** the whitespace check on purpose: the blank-input path must only ever be reached by a genuine string, otherwise a non-string falsy value would take the "omit silently" branch instead of the "reject loudly" one.

Test: `langsmith create feedback rejects non-numeric input types`, parameterised over `false`, `true`, `[]` and `{}`. Confirmed red first — `false` and `true` both produced a score before this change.

For the record this is the third finding on this one small function, all three inherited from the `parseScore` closure I moved out of the block. The move was still the right call — it is what put these paths under test at all — but it is fair to say the original was weaker than it looked.

File: apps/sim/tools/firecrawl/types.ts [hunk 13]
```diff
@@ -123,33 +144,43 @@ export const CRAWLED_PAGE_OUTPUT_PROPERTIES = {
...
+    description: 'Search result title from search engine',
+    optional: true,
+  },
   description: {
     type: 'string',
     description: 'Search result description/snippet from search engine',
+    optional: true,
   },
   url: { type: 'string', description: 'URL of the search result' },
+  position: { type: 'number', description: 'Position of the result', optional: true },
```
[the remark this answers is not in the document]
Comment (waleedlatif1) [reply apps/sim/tools/firecrawl/types.ts:163]: Valid — added in 163f2f4e, with one caveat worth stating plainly because my two sources disagree.

The official SDK types `category?: string` on **both** `SearchResultWeb` and `SearchResultNews`:

```ts
export interface SearchResultWeb  { url: string; title?: string; description?: string; position?: number; category?: string }
export interface SearchResultNews { title?: string; url?: string; snippet?: string; date?: string; imageUrl?: string; position?: number; category?: string }
```

The v2 OpenAPI does **not** list it — I dumped the item property keys to check, and `category` is absent from `web`, `news` and `images` alike. So the schema and the shipped client disagree on this one field.

I have gone with the SDK and declared it `optional: true` on web and news. Reasoning: adding an optional field is additive and cannot break a consumer, whereas omitting a field the provider really sends is precisely the defect this PR exists to fix — and the asymmetry of risk points one way. It is plausibly the echo of the request-side `categories` filter (`github`/`research`/`pdf`/`developer`), which would explain why it is absent from a schema written around the default request.

I did **not** add it to image items: the SDK's `SearchResultImages` does not type it either, so there is no source for it there. A test pins that asymmetry (`leaves category off image items, which the SDK does not type it on`) so it does not get "tidied up" into a symmetry the provider never had.

Tests confirmed red first.

**Comment by waleedlatif1** [comment]
@cubic-dev-ai review

**Comment by waleedlatif1** [comment]
@greptile review

**Comment by waleedlatif1** [comment]
Closing for now — not because of a defect. This batch grew to 17 PRs across ~700 changed call sites, and we would rather revisit it as smaller, independently testable pieces than merge this much at once.

Nothing here is lost: the branch `fix/declared-output-shapes` is preserved and this PR can be reopened. Review state, the reasoning on every thread, and the red-first verification all stay attached.
